File: server/app/api/activities.py
import uuid
from typing import Optional

from fastapi import APIRouter, BackgroundTasks, Depends, File, Form, Query, UploadFile
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.dependencies.auth import get_current_user
from app.dependencies.database import get_db
from app.dependencies.language import get_request_language
from app.models.user import User
from app.schemas.generated import (
    Activity,
    ActivityDetail,
    ActivityResource,
    ActivityResourcePhoto,
    CreateActivityRequest,
    CreateResourceRequest,
    UpdateResourceRequest,
)
from app.services import activity as activity_service
from app.services import activity_resource as resource_service
from app.services.localization import pick

router = APIRouter()


def _activity_schema(a, language: str = "de") -> dict:
    return {
        "id": a.id,
        "title": pick(a.title, a.title_en, language),
        "description": pick(a.description, a.description_en, language),
        "estimated_duration_minutes": a.estimated_duration_minutes,
        "age_min": a.age_min,
        "age_max": a.age_max,
        "cost_indicator": a.cost_indicator,
        "season_relevance": a.season_relevance,
        "weather_suitability": a.weather_suitability,
        "is_partner_content": a.is_partner_content,
        "effort_tier": a.effort_tier,
        "language": a.language,
    }


@router.get("", response_model=list[Activity])
async def list_activities(
    age: Optional[int] = Query(None, ge=0, le=99, description="Child age in years"),
    season: Optional[str] = Query(None, pattern="^(spring|summer|autumn|winter)$"),
    weather: Optional[str] = Query(None, pattern="^(sunny|cloudy|rainy|any)$"),
    cost: Optional[str] = Query(None, pattern="^(free|low_cost)$"),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
    language: str = Depends(get_request_language),
) -> list[dict]:
    activities = await activity_service.list_activities(
        session, user_id=current_user.id, age=age, season=season, weather=weather, cost=cost
    )
    return [_activity_schema(a, language) for a in activities]


@router.post("", response_model=Activity, status_code=201)
async def create_activity(
    body: CreateActivityRequest,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    activity = await activity_service.create_activity(
        session, current_user.id, body, language=current_user.preferred_language
    )
    return _activity_schema(activity)


@router.get("/suggestions", response_model=Activity)
async def get_suggestion(
    child_id: Optional[uuid.UUID] = Query(None, description="Derive age and interests from this child profile"),
    city: Optional[str] = Query(None, description="City name for weather-based suggestions (reserved for future use)"),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
    language: str = Depends(get_request_language),
) -> dict:
    activity = await activity_service.get_suggestion(
        session,
        user_id=current_user.id,
        child_id=child_id,
    )
    return _activity_schema(activity, language)


# ── Activity resources ───────────────────────────────────────────


@router.get("/{activity_id}", response_model=ActivityDetail)
async def get_activity_detail(
    activity_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
    language: str = Depends(get_request_language),
) -> dict:
    activity, can_edit, resources = await resource_service.get_activity_with_resources(
        session, current_user.id, activity_id
    )
    data = _activity_schema(activity, language)
    data["can_edit"] = can_edit
    data["resources"] = resources
    return data


@router.post("/{activity_id}/resources", response_model=ActivityResource, status_code=201)
async def create_activity_resource(
    activity_id: uuid.UUID,
    body: CreateResourceRequest,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    if body.kind == "external":
        return await resource_service.create_external_resource(
            session, current_user.id, activity_id, url=body.url, label=body.label
        )
    return await resource_service.create_internal_text_resource(
        session, current_user.id, activity_id, note_text=body.note_text, label=body.label
    )


@router.post("/{activity_id}/resources/photos", response_model=ActivityResource, status_code=202)
async def create_activity_resource_photo(
    activity_id: uuid.UUID,
    background_tasks: BackgroundTasks,
    image: UploadFile = File(...),
    note_text: Optional[str] = Form(None),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    photo_data = await image.read()
    resource, raw_key, final_key, photo_id = await resource_service.create_photo_only_resource(
        session,
        current_user.id,
        activity_id,
        photo_data=photo_data,
        content_type=image.content_type or "image/jpeg",
        note_text=note_text,
    )
    background_tasks.add_task(
        resource_service.compress_resource_photo, photo_id, raw_key, final_key, settings.DATABASE_URL
    )
    return resource


@router.post(
    "/{activity_id}/resources/{resource_id}/photos",
    response_model=ActivityResourcePhoto,
    status_code=202,
)
async def add_activity_resource_photo(
    activity_id: uuid.UUID,
    resource_id: uuid.UUID,
    background_tasks: BackgroundTasks,
    image: UploadFile = File(...),
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    photo_data = await image.read()
    photo, raw_key, final_key, photo_id = await resource_service.add_photo(
        session,
        current_user.id,
        activity_id,
        resource_id,
        photo_data=photo_data,
        content_type=image.content_type or "image/jpeg",
    )
    background_tasks.add_task(
        resource_service.compress_resource_photo, photo_id, raw_key, final_key, settings.DATABASE_URL
    )
    return photo


@router.patch("/{activity_id}/resources/{resource_id}", response_model=ActivityResource)
async def update_activity_resource(
    activity_id: uuid.UUID,
    resource_id: uuid.UUID,
    body: UpdateResourceRequest,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> dict:
    return await resource_service.update_resource(
        session,
        current_user.id,
        activity_id,
        resource_id,
        label=body.label,
        url=body.url,
        note_text=body.note_text,
    )


@router.delete("/{activity_id}/resources/{resource_id}", status_code=204)
async def delete_activity_resource(
    activity_id: uuid.UUID,
    resource_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> None:
    await resource_service.delete_resource(session, current_user.id, activity_id, resource_id)


@router.delete("/{activity_id}/resources/{resource_id}/photos/{photo_id}", status_code=204)
async def delete_activity_resource_photo(
    activity_id: uuid.UUID,
    resource_id: uuid.UUID,
    photo_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> None:
    await resource_service.delete_photo(session, current_user.id, activity_id, resource_id, photo_id)

File: client/lib/api/activities.ts
import { Platform } from 'react-native';
import { apiClient } from './client';

export interface ActivityItem {
  id: string;
  title: string;
  description: string;
  estimated_duration_minutes: number;
  age_min: number;
  age_max: number;
  cost_indicator: 'free' | 'low_cost' | 'paid';
  season_relevance: string[] | null;
  weather_suitability: string[] | null;
  is_partner_content: boolean;
  effort_tier?: 'casual' | 'dedicated';
}

export interface ActivityFilters {
  age?: number;
  season?: string;
  weather?: string;
  cost?: string;
}

export interface CreateActivityPayload {
  title: string;
  description?: string | null;
  estimated_duration_minutes?: number | null;
  cost_indicator?: 'free' | 'low_cost';
}

export type ResourceKind = 'external' | 'internal';

export interface ActivityResourcePhoto {
  id: string;
  status: 'processing' | 'ready';
  position: number;
  photo_url?: string | null;
}

export interface ActivityResource {
  id: string;
  kind: ResourceKind;
  position: number;
  label?: string | null;
  url?: string | null;
  note_text?: string | null;
  photos?: ActivityResourcePhoto[] | null;
}

export interface ActivityDetail extends ActivityItem {
  can_edit: boolean;
  resources: ActivityResource[];
}

export interface CreateResourcePayload {
  kind: ResourceKind;
  label?: string | null;
  url?: string | null;
  note_text?: string | null;
}

export interface UpdateResourcePayload {
  label?: string | null;
  url?: string | null;
  note_text?: string | null;
}

/** Appends an image to a FormData in a way that works on both web and native. */
async function appendImage(form: FormData, imageUri: string, mimeType: string) {
  if (Platform.OS === 'web') {
    const res = await fetch(imageUri);
    const blob = await res.blob();
    form.append('image', blob, 'photo.jpg');
  } else {
    form.append('image', { uri: imageUri, type: mimeType, name: 'photo.jpg' } as any);
  }
}

export const activitiesApi = {
  list: (filters: ActivityFilters = {}) => {
    const params = Object.fromEntries(
      Object.entries(filters).filter(([, v]) => v !== undefined && v !== null)
    );
    return apiClient.get<ActivityItem[]>('/activities', { params });
  },

  create: (payload: CreateActivityPayload) => apiClient.post<ActivityItem>('/activities', payload),

  suggestion: (childId?: string | null, city?: string | null) => {
    const params: Record<string, string> = {};
    if (childId) params.child_id = childId;
    if (city) params.city = city;
    return apiClient.get<ActivityItem>('/activities/suggestions', { params });
  },

  getDetail: (activityId: string) =>
    apiClient.get<ActivityDetail>(`/activities/${activityId}`),

  createResource: (activityId: string, payload: CreateResourcePayload) =>
    apiClient.post<ActivityResource>(`/activities/${activityId}/resources`, payload),

  createResourcePhoto: async (activityId: string, imageUri: string, mimeType: string, noteText?: string | null) => {
    const form = new FormData();
    await appendImage(form, imageUri, mimeType);
    if (noteText) form.append('note_text', noteText);
    return apiClient.post<ActivityResource>(`/activities/${activityId}/resources/photos`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
  },

  addResourcePhoto: async (activityId: string, resourceId: string, imageUri: string, mimeType: string) => {
    const form = new FormData();
    await appendImage(form, imageUri, mimeType);
    return apiClient.post<ActivityResourcePhoto>(
      `/activities/${activityId}/resources/${resourceId}/photos`,
      form,
      { headers: { 'Content-Type': 'multipart/form-data' } }
    );
  },

  updateResource: (activityId: string, resourceId: string, payload: UpdateResourcePayload) =>
    apiClient.patch<ActivityResource>(`/activities/${activityId}/resources/${resourceId}`, payload),

  deleteResource: (activityId: string, resourceId: string) =>
    apiClient.delete(`/activities/${activityId}/resources/${resourceId}`),

  deleteResourcePhoto: (activityId: string, resourceId: string, photoId: string) =>
    apiClient.delete(`/activities/${activityId}/resources/${resourceId}/photos/${photoId}`),
};
